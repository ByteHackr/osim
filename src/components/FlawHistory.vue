<script setup lang="ts">
import { computed, ref } from 'vue';

import { usePagination } from '@/composables/usePagination';

import { capitalize, formatDateWithTimezone } from '@/utils/helpers';
import { flawFieldNamesMapping } from '@/constants/flawFields';
import type { ZodFlawHistoryItemType } from '@/types/zodFlaw';

import LabelCollapsible from './widgets/LabelCollapsible.vue';

const props = defineProps<{
  history: null | undefined | ZodFlawHistoryItemType[];
}>();

const historyExpanded = ref(true);

const itemsPerPage = 20;
const totalPages = computed(() =>
  Math.ceil((props.history?.length || 0) / itemsPerPage),
);

const {
  changePage,
  currentPage,
  pages,
} = usePagination(totalPages, itemsPerPage);

const paginatedHistoryItems = computed(() => {
  const start = (currentPage.value - 1) * itemsPerPage;
  const end = start + itemsPerPage;
  return props.history?.slice(start, end);
});

function isDateField(field: string) {
  return field.includes('_dt');
}
</script>

<template>
  <LabelCollapsible
    class="my-2"
    :class="{'pb-4': !history?.length}"
    :isExpanded="historyExpanded"
    @toggleExpanded="historyExpanded = !historyExpanded"
  >
    <template #label>
      <label class="mx-2 mb-0 form-label">
        <h4 :class="{'mb-0': !historyExpanded}">History</h4>
      </label>
    </template>
    <template v-if="!history?.length">
      <span>There are no tracked changes for this flaw.</span>
    </template>
    <div v-else class="mt-2">
      <label class="mx-2 form-label w-100">
        <template v-for="historyEntry in paginatedHistoryItems" :key="historyEntry.pgh_slug">
          <div v-if="historyEntry.pgh_diff" class="alert alert-info mb-1 p-2">
            <span>
              {{ formatDateWithTimezone(historyEntry.pgh_created_at || '') }}
              - {{ historyEntry.pgh_context?.user || 'System' }}
            </span>
            <ul class="mb-2">
              <li v-for="(diffEntry, diffKey) in historyEntry.pgh_diff" :key="diffKey">
                <div class="ms-3 pb-0">
                  <span>{{ capitalize(historyEntry.pgh_label) }}</span>
                  <span class="fw-bold">{{ ' ' + (flawFieldNamesMapping[diffKey] || diffKey) }}</span>
                  <span>{{ ': ' +
                    (isDateField(diffKey) && diffEntry[0]
                      ? formatDateWithTimezone(diffEntry[0])
                      : (diffEntry[0]?.toString() || '')
                    ) + ' '
                  }}</span>
                  <i class="bi bi-arrow-right" />
                  {{ (isDateField(diffKey) && diffEntry[1]
                    ? formatDateWithTimezone(diffEntry[1])
                    : (diffEntry[1]?.toString() || '')
                  ) }}
                </div>
              </li>
            </ul>
          </div>
        </template>
      </label>
    </div>
  </LabelCollapsible>
  <div v-if="history?.length" class="pagination-controls gap-1 my-2">
    <button
      type="button"
      tabindex="-1"
      class="btn btn-sm btn-secondary rounded-end-0"
      :disabled="currentPage === 1"
      @click="changePage(currentPage - 1)"
    >
      <i class="bi bi-arrow-left fs-5" />
    </button>
    <button
      v-for="page in pages"
      :key="page"
      tabindex="-1"
      class="osim-page-btn btn btn-sm rounded-0 btn-secondary"
      style="width: 34.8px;"
      :disabled="page === currentPage || page === '..'"
      @click.prevent="changePage(page as number)"
    >
      {{ page }}
    </button>
    <button
      type="button"
      tabindex="-1"
      class="btn btn-sm btn-secondary rounded-start-0"
      :disabled="currentPage === totalPages || totalPages === 0"
      @click.prevent="changePage(currentPage + 1)"
    >
      <i class="bi bi-arrow-right fs-5" />
    </button>
  </div>
</template>

<style scoped lang="scss">
    .pagination-controls {
  display: flex;

  button {
    height: 2rem;
    padding-block: 0;

    &.osim-page-btn:disabled {
      background-color: transparent;
      color: black;
    }
  }
}
</style>
