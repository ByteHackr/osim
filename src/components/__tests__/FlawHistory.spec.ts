import { mount } from '@vue/test-utils';

import FlawHistory from '@/components/FlawHistory.vue';

import type { ZodFlawHistoryItemType } from '@/types/zodFlaw';

import { osimEmptyFlawTest, osimFullFlawTest } from './test-suite-helpers';

function sampleHistoryItem(): ZodFlawHistoryItemType {
  return {
    pgh_created_at: '2024-10-04T15:06:56.760289Z',
    pgh_slug: 'osidb.FlawAudit:123456',
    pgh_label: 'update',
    pgh_context: { url: 'osidb/api/v1/flaws/12d3820a-a43b-417c-a22e-46e47c232a63', user: 1 },
    pgh_diff: { owner: ['', 'noemail@example.com'] },
  };
}

describe('flawHistory', () => {
  osimEmptyFlawTest('is not shown if no history present on flaw', async ({ flaw }) => {
    const subject = mount(FlawHistory, {
      props: {
        history: flaw.history,
      },
    });
    const historyItems = subject.findAll('.alert-info');
    expect(historyItems.length).toBe(0);
  });

  osimFullFlawTest('is shown if history present on flaw', async ({ flaw }) => {
    flaw.history = [];
    flaw.history.push(sampleHistoryItem());
    const subject = mount(FlawHistory, {
      props: {
        history: flaw.history,
      },
    });
    const historyItems = subject.findAll('.alert-info');
    expect(historyItems.length).toBe(1);
    expect(subject.html()).toMatchSnapshot();
  });

  osimFullFlawTest('displays history items', async ({ flaw }) => {
    flaw.history = [];
    flaw.history.push(sampleHistoryItem());
    const subject = mount(FlawHistory, {
      props: {
        history: flaw.history,
      },
    });
    const historyListItem = subject.find('li div');
    expect(historyListItem?.text()).includes('Update Owner:');
  });
});
