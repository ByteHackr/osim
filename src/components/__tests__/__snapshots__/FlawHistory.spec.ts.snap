// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`flawHistory > is shown if history present on flaw 1`] = `
"<div data-v-357e30d7="" class="osim-collapsible-label my-2 my-2"><button data-v-357e30d7="" type="button" class="me-2 osim-collapsible-toggle"><i data-v-357e30d7="" class="bi me-1 bi-plus-square-dotted"></i><label class="mx-2 mb-0 form-label">
      <h4 class="mb-4">History</h4>
    </label></button>
  <div data-v-357e30d7="" class="ps-3 border-start">
    <div data-v-357e30d7="" class="visually-hidden">
      <div class="mt-2"><label class="mx-2 form-label w-100">
          <div class="alert alert-info mb-1 p-2"><span>2024-10-04 15:06 UTC - 1</span>
            <ul class="mb-2">
              <li>
                <div class="ms-3 pb-0"><span>Update</span><span class="fw-bold"> Owner</span><span>:  </span><i class="bi bi-arrow-right"></i> noemail@example.com</div>
              </li>
            </ul>
          </div>
        </label></div>
    </div>
  </div>
</div>"
`;
